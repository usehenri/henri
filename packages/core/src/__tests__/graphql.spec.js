const BaseModule = require('../base/module');
const Henri = require('../henri');
const Graphql = require('../1.graphql');

describe('graphql', () => {
  beforeAll(async () => {
    this.henri = new Henri({ runlevel: 1 });
    await this.henri.init();
  });

  afterAll(async () => {
    await this.henri.stop();
  });

  test('should be defined', () => {
    expect(this.henri.graphql).toBeDefined();
  });

  test('should extend BaseModule', () => {
    expect(this.henri.graphql).toBeInstanceOf(BaseModule);
  });

  test('should match snapshot', () => {
    const gql = new Graphql();
    expect(gql).toMatchSnapshot();
  });

  test('should load a new endpoint', () => {
    expect(this.henri.graphql.endpoint).toEqual('/_henri/graph');
  });

  test('should register middlewares', () => {
    expect(this.henri._middlewares).toHaveLength(2);

    const mockApp = {
      use: jest.fn(),
    };

    for (let middleware of this.henri._middlewares) {
      expect(typeof middleware).toEqual('function');
      middleware(mockApp);
    }

    expect(mockApp.use).toHaveBeenCalledTimes(2);
  });

  describe('extract', () => {
    test('should have extract function', () => {
      expect(typeof this.henri.graphql.extract).toEqual('function');
    });

    test('should return false on empty models', () => {
      expect(this.henri.graphql.extract({ abc: 'd' })).toBeFalsy();
    });

    test('should extract from model', () => {
      const data = [
        {
          title: 'Le bonheur de vivre',
          year: 1905,
        },
        { title: 'Music', year: 1910 },
      ];

      const model = {
        graphql: {
          types: `
            type Query { artworks: [Artwork], artwork: Artwork }
          `,
          resolvers: { Query: { artworks: () => data } },
        },
      };

      this.henri.graphql.extract(model);
      this.henri.graphql.extract({
        graphql: { types: `type Artwork { title: String, year: Int }` },
      });
      this.henri.graphql.extract({
        graphql: { resolvers: { Query: { artwork: key => data[key] } } },
      });

      expect(this.henri.graphql.typesList).toEqual(
        expect.arrayContaining([model.graphql.types])
      );

      expect(this.henri.graphql.typesList).toEqual(
        expect.arrayContaining([`type Artwork { title: String, year: Int }`])
      );
    });
  });

  describe('merge', () => {
    test('should compile schema', async () => {
      expect(this.henri.graphql.typesList.length).toBeGreaterThan(0);
      expect(this.henri.graphql.schema).toBeNull();
      expect(this.henri.graphql.active).toBeFalsy();

      this.henri.graphql.merge();
      expect(this.henri.graphql.schema).toBeTruthy();
      expect(this.henri.graphql.active).toBeTruthy();

      let result = await this.henri.graphql.run(`{ artworks { title, year }}`);
      expect(result).toMatchSnapshot();
    });

    test('should clear on reload', async () => {
      this.henri.graphql.reload();

      expect(this.henri.graphql.schema).toBeNull();
      expect(this.henri.graphql.active).toBeFalsy();

      this.henri.graphql.merge();

      expect(this.henri.graphql.active).toBeFalsy();

      const result = await this.henri.graphql.run();
      expect(result).toEqual('No graphql schema found.');
    });
  });
});
