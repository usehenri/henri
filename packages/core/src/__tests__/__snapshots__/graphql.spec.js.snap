// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql merge should compile schema 1`] = `
Object {
  "data": Object {
    "artworks": Array [
      Object {
        "title": "Le bonheur de vivre",
        "year": 1905,
      },
      Object {
        "title": "Music",
        "year": 1910,
      },
    ],
  },
}
`;

exports[`graphql should match snapshot 1`] = `
Graphql {
  "active": false,
  "consoleOnly": false,
  "endpoint": "/_henri/gql",
  "extract": [Function],
  "henri": null,
  "init": [Function],
  "key": null,
  "merge": [Function],
  "name": "graphql",
  "reload": [Function],
  "reloadable": true,
  "resolvers": null,
  "resolversList": Array [],
  "run": [Function],
  "runlevel": 1,
  "schema": null,
  "types": null,
  "typesList": Array [],
}
`;
